### ISSUE TITLE
Workaround for running on ROCm platform?

### ISSUE THREAD

**Issue by SparkJiao** [OpenRLHF/OpenRLHF#667]
I encouter the following problem when running on 8*MI300x with ROCM==6.1.
<img width="2034" alt="image" src="https://github.com/user-attachments/assets/74b41794-caae-4ada-84ab-caffff92935c" />

I saw several merge commits: #536 ; and opening issue in Ray [#49148](https://github.com/ray-project/ray/pull/49148).

May I confirm is currently this problem solved or any workaround?

Thank you very much!

BTW, I'm using `torch==2.5.1, ray=2.12.0, deepspeed==0.15.1`.

**Comment by hijkzzz** [comment]
# For AMD GPUs:
export RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES=1
--vllm_sync_backend gloo

**Comment by SparkJiao** [comment]
It seems that this workaround does not work. Do you have any other suggestions?

**Comment by HollowMan6** [comment]
> It seems that this workaround does not work.

@SparkJiao Do you still have the same tuple index out of the range traceback even after you set `RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES=1`? Please make sure you set `RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES` before you start the ray (`ray start ...`), or a cleaner way is to use:

```bash
ray job submit \
--runtime-env-json='{"env_vars": {"RAY_EXPERIMENTAL_NOSET_ROCR_VISIBLE_DEVICES": "1"}}' \
...
```

when you submit your job.

**Comment by HollowMan6** [comment]
@SparkJiao If it still doesn't work, you can apply my patch to ray https://github.com/ray-project/ray/pull/49148/files anyway (Just remove `AMDGPUAcceleratorManager` so that it goes to `NvidiaGPUAcceleratorManager` for AMD GPUs in `python/ray/_private/accelerators/__init__.py`).

The final step if none of the above works is to downgrade the ray version to `2.9.3`.

If you have any other questions related to AMD, for easier discussion, feel free to send me your WeChat account through my email listed on my GitHub profile.

**Comment by SparkJiao** [comment]
@HollowMan6 Thank you for your kind help very much!

I have put this env variable between `ray start --head ...` and `ray job submit ...`. I will update the script now and try again.

**Comment by SparkJiao** [comment]
@HollowMan6 It works. Thank you very much!
Also appreciate it to @hijkzzz !

**Comment by HollowMan6** [comment]
For people who find this issue later, if you don't want to set `--vllm_sync_backend gloo` and still prefer `--vllm_sync_backend nccl`, you will need to disable intra-node communication with

```bash
export NCCL_P2P_LEVEL=LOC
export NCCL_SHM_DISABLE=1
```

The reason seems to be that for AMD platforms, ROCm NCCL (RCCL) has some issues finding the correct communication path (mixing P2P with GDRDMA) when the communication topology is not symmetrical (e.g. 1 GPU on 1 node communicates with 8 GPUs on the other node), while in OpenRLHF, the vllm weight update topology is unsymmetrical (actor rank 0 on one node communicates with all vLLM engines on other nodes)

**Comment by Treemann** [comment]
Hi @HollowMan6, I find some issues running on AMD's MI300 series GPU and wonder if you have met that before. I've sent the details to your mailbox (hollowman@opensuse.org). Could you please check? Thanks!

**Comment by HollowMan6** [comment]
Hi @Treemann , just found possibly your email (?) in the spam folder, sorry about that:

> Issues like GPU hang and loss explosion occur occasionally on AMD MI300 series GPU while everything works well on NVIDIA GPU. 
Since you have run OpenRLHF on AMD MI300 GPU, I wonder if you have also verified the accuracy or met issues similar to us. 

Unfortunately, I was using MI250X instead of MI300 on my side. For the GPU hang issue, is it caused by NCCL timeout or GPU OOM? I'm not sure about it then if you have already tried this: https://github.com/OpenRLHF/OpenRLHF/issues/667#issuecomment-2708870780 Most likely these are caused by your software toolchain versions somehow, and I would encourage you to debug these things.

For accuracy, I didn't compare AMD directly with NVIDIA, but it can reach convergence after some hyper-parameter fine-tuning for my use case. If you compare them directly and that happens, my guess is that it's caused by the precision difference between different hardware.

**Comment by Treemann** [comment]
@HollowMan6, I have tried different solutions and the GPU hang issue finally solved. I found the precision difference really affected the outputs but not sure if there is any other bug behind that. Will try hyper-parameter fune-tuning. Thanks for your reply!
